### ISSUE TITLE
Allow plugins to override `DefaultPolicy`

### ISSUE THREAD

**Issue by gmazzap** [composer/composer#9362]
Composer v2 `PluginEvents::PRE_POOL_CREATE` makes it possible to filter the packages to be used in the `Pool`.

If that provides a degree of power to plugins, it could be possible to do very interesting things if plugins would be to control how filtered packages are "solved" and being able to provide a custom Policy is probably the simplest approach, considering that a `PolicyInterface` already exists with a `DefaultPolicy` implementation that has nor alternatives.

The point is that even if the "available" packages can be customized, the customization possibility is restricted by the "hardcoded" requirements: changing packages to something that doesn't pass current requirements (defined in files very often controlled by third parties) will make the installation fails.

Besides the `Policy`, to provide complete control to plugins, it would be also necessary to allow overriding `Request` requirements, something that is right now not possible (unless using one of the "tricks" PHP provide to write in private variables): https://github.com/composer/composer/blob/master/src/Composer/DependencyResolver/Request.php#L64-L66

I have a concrete usage example, but it is not really possible to explain it shortly. So the rest of the text below is for those interested in knowing more about the specific use case.

---

## Context

Since I'm using Composer, I've always found the _local_ development of libraries intended to be part of a framework (or alike) to be problematic.

Imagine this scenario:

- there's a framework and a module/plugin/bundle for it. This module exists a remote Composer repo, e.g. Packagist
- the framework has a mechanism to install modules/plugins/bundles based on Composer
- a module developer needs to work on it. Because the framework is based on Composer, if the module is installed in the "normal" way, it would be pulled from the Composer repository, not from the locally modified package, so it is not possible to test locally the module/plugin/bundle in the context of the framework.

There're different ways to deal with this case. One of the most effective ways I've found is to have a separate `composer.json`, e. g. `composer.local.json` where the modules/plugins/bundles to develop are required using `"@dev"` and path repositories exist pointing each module/plugin/bundle to develop. After that, updating dependencies like `COMPOSER=composer.local.json composer update` everything works as expected.

This workflow is not perfect. Among other things:
- the command to execute is quite long
- the `composer.local.json` must be copied _entirely_. That means that if a developer wants to keep it, it has to be kept in line with the `composer.json` and this becomes quite hard, especially when multiple people work on the project. When I use this approach, I usually delete the `composer.local.json` when done with changes and then create again copying from `composer.json` when I need to develop something else. This _kind of_ works (not really well when multiple people work on multiple modules/plugins/bundles at the same time), but surely is not an ideal development experience.

## Studio

Another quite effective solution I used in the past, is to have a separate local-only branch in my project where there's a `composer.json` that requires all the packages I might need to work on using `@dev` and a `studio.json` to support ["Studio"](https://github.com/franzliedke/studio).

This approach works quite well because I can use the usual `composer update` command, and I don't need to maintain two separate `composer.json`, and considering Studio supports glob patterns for the local paths to load, normally my `studio.json` files are a couple of lines long.

However, Studio always had three issues for me:

- it does not work with the `require` command (edge case, maybe)
- there's no way to decide when to use path repo and when to use normal repo, as soon the `studio.json` it is used (this is why I use a separate branch to have `studio.json`)
- not a Studio issue, but using path repositories it is subject to #8389

But the biggest problem, which right now makes it for me unusable, is that it does not support Composer 2, and it seems that the workflow it implements can not really work in Composer 2. See https://github.com/franzliedke/studio/pull/104#issuecomment-648112556

## A possible new plugin

Because of the impossibility to use Studio, I thought about creating my own plugin, that would work similarly to Studio but without its issues and with support for (only) Composer v2.

My idea was to:

- Load a separate JSON file (similarly to `studio.json`) that contains a list of local `composer.json` files. For each, create a Package instance, hardcoding the version to `dev-local`. Group these packages in an array repository to be prepended in the `RepositoryManager`.
- Make use of `PluginEvents::PRE_POOL_CREATE` to filter the `Pool` packages, to make sure local packages are present only once, with the `dev-local` constraint
- If "local packages" are required in the root, override the constraint in the Request to be `dev-local` (to pass `Solver::checkForRootRequireProblems`)
- Create a custom `Policy` implementation, aware of the local packages that will make the Solver accept `dev-local` for them, no matter the actual requirement

My idea was also that when `--no-dev` flag is used, _nothing_ of the above would happen. After all, this workflow is for local _development_.

This would make it possible to have a single `composer.json`, a single branch, and to don't have to worry about the requirements. 

In the implementation points list above, the first two items are possible. The third is possible via a "hackish" solution, and the fourth is not possible. Hence this issue.

**Comment by naderman** [comment]
To focus on your use case first, what problems do you see with usind "--prefer-source" when installing the framework's deps, and then working on your plugin/module/bundle inside the git clone you receive inside your vendor directory?

**Comment by Seldaek** [comment]
https://github.com/composer/composer/issues/601 is our 2nd oldest issue still open.. and IMO it would be good to add support for this natively in Composer somehow, especially if studio is broken beyond repair with v2.

It'd be good to get a list of requirements though to make sure we build something which works for the people having this use case. If you like to help implementing it in Composer that might be a thing too for sure, but I'd really like to first discuss *how* we are going to approach this.

**Comment by stof** [comment]
>     * Create a custom `Policy` implementation, aware of the local packages that will make the Solver accept `dev-local` for them, no matter the actual requirement

This is not something that a custom policy can do. The Policy is about selecting the preferred package among a list, but it cannot bypass the solver.

**Comment by gmazzap** [comment]
@naderman I've experienced that more often than not I need a "disposable" vendor folder. When something goes wrong I can often solve the issue by deleting the lock and the vendor folder and do an installation from scratch. That is especially true when there are plugins that mess with the files generated by Composer. This is why I prefer to have my Git sources outside another package's vendor. On top of that, `--prefer-source` apply to _all_ packages. The last project I'm working on has ~100 packages and the state I get for all of them when using `--prefer-source` is different from what I get when not using that. So using this approach the tests I conduct locally will be less reliable. 

@Seldaek I'm surely happy to help in having this in Composer. Are you still convinced about using the npm link feature as a base? How do you think we should move this forward? Do you want to propose something I can work on? Do you want me to come up with a proposal to discuss? If this feature is embedded in Composer I'm sure there's a less convoluted way to make it work compared to what a plugin can do. Do you think it is worth opening a separate issue for the discussion? Or use this issue, or maybe #601? Something else?

@stof If this becomes a core feature I'm sure there will be better solutions. That said, I don't want to bypass the Solver. I've not done any test in this direction, but considering that the Pool is edited to contain a single reference to a package _x/y_ in the version _dev-local_, I thought that having a completely custom Policy would not be enough to make sure that whatever version is required of that package the Solver would consider _dev-local_ good to satisfy the requirement. Isn't that true? Is the list passed to the policy a list of packages that _passed_ the requirements? I planned to dig a bit more into how the whole workflow works, but having some direction from someone who knows already could be helpful...

**Comment by naderman** [comment]
> @naderman I've experienced that more often than not I need a "disposable" vendor folder. When something goes wrong I can often solve the issue by deleting the lock and the vendor folder and do an installation from scratch. That is especially true when there are plugins that mess with the files generated by Composer. This is why I prefer to have my Git sources outside another package's vendor. On top of that, `--prefer-source` apply to _all_ packages. The last project I'm working on has ~100 packages and the state I get for all of them when using `--prefer-source` is different from what I get when not using that. So using this approach the tests I conduct locally will be less reliable.

You can select source install for specific packages only with preferred-install: https://getcomposer.org/doc/06-config.md#preferred-install

However this all won't help if you want to use the same local directory in multiple projects.


> @stof If this becomes a core feature I'm sure there will be better solutions. That said, I don't want to bypass the Solver. I've not done any test in this direction, but considering that the Pool is edited to contain a single reference to a package _x/y_ in the version _dev-local_, I thought that having a completely custom Policy would not be enough to make sure that whatever version is required of that package the Solver would consider _dev-local_ good to satisfy the requirement. Isn't that true? Is the list passed to the policy a list of packages that _passed_ the requirements? I planned to dig a bit more into how the whole workflow works, but having some direction from someone who knows already could be helpful...

That's why I just wanted to focus on the use case for now. The policy definitely won't help with this. It only selects a preferred package from a list of free choices when all of them could match requirements. It often won't even get called if requirements dictate what to choose, so you could still end up with other versions installed it they aren't entirely removed from the pool, and if you do that you don't need to change the policy.

**Comment by gmazzap** [comment]
> You can select source install for specific packages
But this means I have to change the `composer.json`, something I don't want to do in the project deployed online, meaning that I need to use a separate `composer.json`, doable, but not best experience @naderman 

> The policy definitely won't help with this. 
Thank you for explaining. If there's the will to have a feature native to Composer, a solution can be found without in a less convoluted way, and I'm here to help. But, as @Seldaek said, before I would start writing any code it is important we agree on a line I can follow.

**Comment by naderman** [comment]
> > You can select source install for specific packages
> But this means I have to change the `composer.json`, something I don't want to do in the project deployed online, meaning that I need to use a separate `composer.json`, doable, but not best experience @naderman

Usually that isn't a problem, as you would set preferred-install for dev installation and your prod deployments would use --prefer-dist which would override the preferred setting from your composer.json.
 
> > The policy definitely won't help with this.
> Thank you for explaining. If there's the will to have a feature native to Composer, a solution can be found without in a less convoluted way, and I'm here to help. But, as @Seldaek said, before I would start writing any code it is important we agree on a line I can follow.
Indeed, all for that, but I don't have a clear idea of how this should work yet either.

**Comment by gmazzap** [comment]
@naderman @Seldaek I spend some time in shaping an idea on how this could work. Where do I start a conversation?

This is what I have so far: https://gist.github.com/gmazzap/b71e0d4eb99e0657d939bbc8d856bc7d 

Let me know if that sounds reasonable to you and if/how we can move the conversation forward.

**Comment by Sweetchuck** [comment]
I accidentally run into this issue, because I am looking for another one.

For the „_... here's a framework and a module/plugin/bundle ..._” problem I come up with a sort of solution.

* https://github.com/Sweetchuck/composer-suite
* optional https://github.com/Sweetchuck/composer-repo-path

I use them on several project.

**Comment by Seldaek** [comment]
Kinda feel like workarounds have been found in various plugins so maybe we can just close this here.
